odoo/addons/base/models/ir_actions_report.py

```diff
@@ -785,10 +786,15 @@ def _get_unreadable_pdfs(self, streams):
             result_stream = io.BytesIO()
             try:
                 reader = PdfFileReader(stream)
+                if reader.isEncrypted:
+                    # If the PDF is owner-encrypted, try to unwrap it by giving it an empty user password.
+                    reader.decrypt('')
                 writer.appendPagesFromReader(reader)
                 writer.write(result_stream)
             except utils.PdfReadError:
                 unreadable_streams.append(stream)
+            except (NotImplementedError, StructError) as e:
+                _logger.warning("Unable to access the attachments of %s. Tried to decrypt it, but %s.", self, e)
```

Comment:
I agree, but in this case with this pdf, if we don't rise the error, we are able to generate the pdf without any problem, this is a limitation on PyPDF2 level, that's why I think if it's just `NotImplementedError` exception we ignore it and continue, I'm not sure if it's the right decision, but that's the idea we agreed to use on this fix https://github.com/odoo/odoo/commit/c4ae016d4485501fea888d3e4533be71f599a81d which is similar to this issue

Reply:
@odony the problem is that when calling `appendPagesFromReader()` PyPDF2 will try to decrypt the PDF, and when it fails it raise `PdfReadError` instead of `NotImplementedError` which gives the impression that we can't read the PDF 
```py
    def getNumPages(self):
        if self.isEncrypted:
            try:
                self._override_encryption = True
                self.decrypt('')
                return self.trailer["/Root"]["/Pages"]["/Count"]
            except:
                raise utils.PdfReadError("File has not been decrypted")
```

Reply:
What I find odd about this issue is that downloading a single unreadable original PDF fails. The whole point of the original fix was that we display the unreadable PDF's to the user so the unreadable ones could still be downloaded individually. If we're only downloading a single file it doesn't make sense to check if it can be merged because there doesn't need to be any merging happening.

@thco-odoo had a very similar ticket recently where there wasn't an encryption issue but the PDF was slightly broken in a way PyPDF2 couldn't deal with. In this case the customer was also unable to download the single PDF.

I just did a quick test with the PDF's from the tickets that led to the original fix. In 14.0 uploading the PDF's through the Documents app on a clean db gave a 500 server error due to a company permission issue. That's new. On 15.0 I was able to upload a PDF through Documents and create a vendor bill from it. When I tried to download it next, I got this:
![Screenshot from 2023-03-21 09-02-16](https://user-images.githubusercontent.com/117735233/226548960-6fa509ce-7cd5-4ad0-b98a-8c43fc43ead8.png)

To me this suggests the presence of a regression. I also noticed that in the ticket 3221796 linked to this PR in 14.0 it fails in `_raise_on_unreadable_pdfs` while in later versions it happens in a different spot, like in the screenshot. I'm guessing some code got added in the meanwhile that works fine with PDF's PyPDF2 can read, but that gets called earlier than `_raise_on_unreadable_pdfs` and that also affects single PDF's that don't need to be merged. I suspect the solution is going to lie in addressing the regression. Comparing with the version at the time of the original implementation will probably give a hint on what happened.

@abla001 I can share links to the old PDF's if you like.

Reply:
@mege-odoo I'm really lost here, because every time we got a ticket where PyPDF2 fails on reading a new PDF, I don't really have any idea about a good fix here, if there any suggestion I'm taking it :pray:

Reply:
@abla001 In general I see two strategies for that: mitigation and fixing the root cause. `_raise_on_unreadable_pdfs` is mitigation because at the time PyPDF2 was unmaintained and in stable we're also stuck on the versions that come with the stable Linux distributions that Odoo release supports. Currently, PyPDF2 has seen further maintenance until the end of last year and its successor pypdf appears actively maintained: https://pypi.org/project/PyPDF2/ , https://pypi.org/project/pypdf/ . At the time of the fix I was also looking at pikepdf which seemed more mature and was also actively maintained: https://pypi.org/project/pikepdf/ . Switching to either of these in the next Odoo version could help address the root issue of the PyPDF2 version we use not supporting certain encryption methods and not being fault-tolerant when a PDF doesn't conform 100% to the standard.

For bug fixes in stable we have to work with what we have and do mitigation. In the case of the issue from the description, when printing the original vendor bills from multiple files at once, the UserError from `_raise_on_unreadable_pdfs` popping up if one of them can't be read due to unsupported encryption is expected for me, because that's what it's for: so the user can exclude the troublesome PDF from the PDF's to be merged. What isn't expected is that it does this if you're trying to print the original vendor bill for a single invoice. There's no merging or in-depth reading that would need to be done in that case, because you're just returning the file as-is.

If you look at for example https://github.com/odoo/odoo/blob/8cfb5c425721ad907a1999520817ff752b4ea722/odoo/addons/base/models/ir_actions_report.py#L668 , it's already treating the single invoice case as a special one where it doesn't merge. So for the issue with the single encrypted PDF that can't be downloaded, I'd look for a clean way to detect that we're only dealing with a single file in the places where calls might fails or we detect the PDF is unreadable. For example, if `_raise_on_unreadable_pdfs` raises indicating it found unreadable files, the code calling it could catch that and decide whether or not to re-raise depending on the context. At https://github.com/odoo/odoo/blob/8cfb5c425721ad907a1999520817ff752b4ea722/odoo/addons/base/models/ir_actions_report.py#L879 , we'd only be merging the files passed to `_raise_on_unreadable_pdfs`. If it was passed only a single file, we could ignore it being unreadable and just have `_post_pdf` return the original to the user (which I believe at first glance it will do in that case). In the case of the call at https://github.com/odoo/odoo/blob/8cfb5c425721ad907a1999520817ff752b4ea722/odoo/addons/base/models/ir_actions_report.py#L909 , I suppose there's wkhtmltopdf generated content that's going to be merged together with the content passed to `_raise_on_unreadable_pdfs`, so in that case I guess we keep raising the exception like we currently do.

As for the same issue in the later versions: it seems to be trying to add a banner to the original PDF? That's not going to work if we can't read it. I'd skip adding the banner in that case. It's better than not returning anything at all.

Hope this helps :-)

Reply:
@mege-odoo thank you so much for the explanation that's so helpful.
@nle-odoo thank you for feedback I'm closing this PR as it's fixed by yosa